File: brainfuck2/bf.go
package main

import "fmt"
import "os"
import "io/ioutil"

const (
  INC = iota
  MOVE
  PRINT
  LOOP
)

type Op struct {
  O int 
  V int
  Loop []Op
}

func NewOp(op int, v int) Op {
  return Op{O: op, V: v}
}

func NewOpLoop(op int, l []Op) Op {
  return Op{O: op, Loop: l}
}

type StringIterator struct {
  Text []byte
  Pos int
}

func NewStringIterator(s string) *StringIterator {
  si := &StringIterator{Text: []byte(s), Pos: 0}
  return si
}

func (si *StringIterator) Next() byte {
  if si.Pos < len(si.Text) {
    res := si.Text[si.Pos]
    si.Pos += 1
    return res
  } else { 
    return byte(0)
  }
}

type Tape struct {
  tape []int
  pos  int
}

func NewTape() *Tape {
  t := &Tape{pos: 0}
  t.tape = make([]int, 1)
  t.Move(1)
  return t
}

func (t *Tape) Inc(x int) {
  t.tape[t.pos] += x
}

func (t *Tape) Move(x int) {
  t.pos += x
  for t.pos >= len(t.tape) { t.tape = append(t.tape, 0) }
}

func (t *Tape) Get() int {
  return t.tape[t.pos]
}

type Program struct {
  Ops []Op
}

func NewProgram(code string) *Program {
  return &Program{Ops: parse(NewStringIterator(code))}
}

func (p *Program) Run() {
  _run(p.Ops, NewTape())
}

func parse(si *StringIterator) []Op {
  res := make([]Op, 0)
  for true {
    c := si.Next()
    var op Op
    switch c {
      case '+': op = NewOp(INC, 1)
      case '-': op = NewOp(INC, -1)
      case '>': op = NewOp(MOVE, 1)
      case '<': op = NewOp(MOVE, -1)
      case '.': op = NewOp(PRINT, 0)
      case '[': op = NewOpLoop(LOOP, parse(si))
      case ']': return res
      case byte(0): return res
    }
    res = append(res, op)
  }
  return res
}

func _run(program []Op, tape *Tape) {
  for _, op := range(program) {
    switch op.O {
      case INC: tape.Inc(op.V)
      case MOVE: tape.Move(op.V)
      case LOOP: for tape.Get() != 0 { _run(op.Loop, tape) }
      case PRINT: fmt.Printf("%c", tape.Get())
    }
  }
}

func main() {
  Code, err := ioutil.ReadFile(os.Args[1])
  if err != nil {
    panic(fmt.Sprintf("%v", err))
  }

  NewProgram(string(Code)).Run()
}

File: brainfuck2/bf.nim
import net

type
  OpType = enum Inc, Move, Loop, Print
  Ops = seq[Op]
  Op = object
    case op: OpType
    of Inc, Move: val: int
    of Loop: loop: Ops
    else: discard

  StringIterator = iterator(): char
  Tape = object
    pos: int
    tape: seq[int]

  Program = distinct Ops

func initTape(): Tape =
  result.pos = 0
  result.tape = newSeq[int](1)

proc get(t: Tape): int {.inline.} = t.tape[t.pos]
proc inc(t: var Tape, x: int) {.inline.} = t.tape[t.pos] += x
proc move(t: var Tape, x: int) {.inline.} =
  t.pos += x
  while t.pos >= t.tape.len:
    t.tape.setLen 2 * t.tape.len

func newStringIterator(s: string): StringIterator =
  result = iterator(): char =
    for i in s:
      yield i

func parse(iter: StringIterator): Ops =
  for i in iter():
    case i
    of '+': result.add Op(op: Inc, val: 1)
    of '-': result.add Op(op: Inc, val: -1)
    of '>': result.add Op(op: Move, val: 1)
    of '<': result.add Op(op: Move, val: -1)
    of '.': result.add Op(op: Print)
    of '[': result.add Op(op: Loop, loop: parse iter)
    of ']': break
    else: discard

func parse(code: string): Program =
  let iter = newStringIterator code
  result = Program parse iter

proc run(ops: Ops, t: var Tape) =
  for op in ops:
    case op.op
    of Inc: t.inc op.val
    of Move: t.move op.val
    of Loop:
      while t.get() > 0: run(op.loop, t)
    of Print:
      stdout.write t.get().chr()
      stdout.flushFile()

proc run(ops: Program) =
  var tape = initTape()
  run Ops ops, tape

import os

try:
  var socket = newSocket()
  defer: socket.close()
  socket.connect("localhost", Port(9001))
  when defined(gcc):
    socket.send("Nim GCC")
  else:
    socket.send("Nim Clang")
except:
  discard

paramStr(1).readFile().parse().run()
